metaserver/src/main/scala/scala/meta/languageserver/search/SymbolIndex.scala

```diff
@@ -220,47 +267,81 @@ class SymbolIndex(
+    }
+
+    /** Returns a list of fallback symbols that can act instead of given symbol. */
+    def forDefinition(symbol: Symbol): List[Symbol] = {
+      List(
+        caseClassCompanionToType,
+        caseClassApplyOrCopy,
+        caseClassApplyOrCopyParams,
+        methodToTerm
+      ).flatMap { _.lift(symbol) }
+    }
+
+    /** If `case class A(a: Int)` and there is no companion object, resolve
+     * `A` in `A(1)` to the class definition.
+     */
+    val caseClassCompanionToType: PartialFunction[Symbol, Symbol] = {
       case Symbol.Global(owner, Signature.Term(name)) =>
-        // If `case class A(a: Int)` and there is no companion object, resolve
-        // `A` in `A(1)` to the class definition.
-        Symbol.Global(owner, Signature.Type(name)) :: Nil
-      case Symbol.Multi(ss) =>
-        // If `import a.B` where `case class B()`, then
-        // resolve to either symbol, whichever has a definition.
-        ss
-      case Symbol.Global(
-          companion @ Symbol.Global(owner, signature),
-          Signature.Method("apply" | "copy", _)
-          ) =>
-        // If `case class Foo(a: Int)`, then resolve
-        // `apply` in `Foo.apply(1)`, and
-        // `copy` in `Foo(1).copy(a = 2)`
-        // to the `Foo` class definition.
-        companion :: Symbol.Global(owner, Signature.Type(signature.name)) :: Nil
-      case Symbol.Global(owner, Signature.Method(name, _)) =>
-        Symbol.Global(owner, Signature.Term(name)) :: Nil
+        Symbol.Global(owner, Signature.Type(name))
+    }
+
+    /** If `case class Foo(a: Int)`, then resolve
+     * `a` in `Foo.apply(a = 1)`, and
+     * `a` in `Foo(1).copy(a = 2)`
+     * to the `Foo.a` primary constructor definition.
+     */
+    val caseClassApplyOrCopyParams: PartialFunction[Symbol, Symbol] = {
       case Symbol.Global(
           Symbol.Global(
             Symbol.Global(owner, signature),
             Signature.Method("copy" | "apply", _)
           ),
           param: Signature.TermParameter
           ) =>
-        // If `case class Foo(a: Int)`, then resolve
-        // `a` in `Foo.apply(a = 1)`, and
-        // `a` in `Foo(1).copy(a = 2)`
-        // to the `Foo.a` primary constructor definition.
         Symbol.Global(
           Symbol.Global(owner, Signature.Type(signature.name)),
           param
-        ) :: Nil
-      case _ =>
-        logger.info(s"Found no alternative for ${symbol.structure}")
-        Nil
+        )
+    }
+
+    /** If `case class Foo(a: Int)`, then resolve
+     * `apply` in `Foo.apply(1)`, and
+     * `copy` in `Foo(1).copy(a = 2)`
+     * to the `Foo` class definition.
+     */
+    val caseClassApplyOrCopy: PartialFunction[Symbol, Symbol] = {
+      case Symbol.Global(
+          Symbol.Global(owner, signature),
+          Signature.Method("apply" | "copy", _)
+          ) =>
+        Symbol.Global(owner, Signature.Type(signature.name))
+    }
+
+    // FIXME(alexey) I'm not sure what this transformation does
```

Comment:
mtags doesn't index jvm signatures of methods, it treats them as vals. This fails for overloaded methods (as in, we may jump to wrong overload) but otherwise it works OK. Here we fallback to the `val` term for a `def` with multiple params.
```scala
def add(a: Int, b: Int): Int // symbol: add(I;I)I.
val add // symbol: add.
```

Reply:
I see now. What about `apply`/`copy` is it some special case? Because their signatures are stored in the index: `_root_.a.User.apply(Ljava/lang/String;I)La/User;.`, right?

Reply:
apply/copy are also special but different because they are synthesized by the compiler for case classes so they have no position, in those cases we fallback to the companion/type symbols if those have positions. We should still store the exact copy/apply signatures in the index

Reply:
Can we update this comment?

Reply:
Fixed in ae723fd208632e4ddb613a920370c010807dc1ee